offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/dataresources/persons/CustomerDbConnector.groovy

```diff
@@ -274,16 +301,103 @@ class CustomerDbConnector implements CreateCustomerDataSource, UpdateCustomerDat
+      log.error(e.stackTrace.join("\n"))
+      throw new DatabaseQueryException("The customer's affiliations could not be updated.")
+    }
+    
+    List<Affiliation> newAffiliations = new ArrayList<>();
+    
+    // find added affiliations - could use set operations here, but we have lists...
+    for(Affiliation affiliation : updatedAffiliations) {
+      if(!existingAffiliations.contains(affiliation)) {
+        newAffiliations.add(affiliation)
+      }
+    }
+    if(newAffiliations.isEmpty()) {
+      throw new DatabaseQueryException("Customer already has provided affiliation(s), no update was necessary.")
+    }
+
+    Connection connection = connectionProvider.connect()
+    connection.setAutoCommit(false)
+
+    connection.withCloseable {it ->
+      try {
+       
+    storeAffiliation(connection, customerId, newAffiliations)
+    
+      } catch (Exception e) {
+        log.error(e.message)
+        log.error(e.stackTrace.join("\n"))
+        connection.rollback()
+        connection.close()
+        throw new DatabaseQueryException("Could not update customer's affiliations.")
+      }
+    }
+  }
 
+  private void changeCustomerActiveFlag(int customerId, boolean active) {
+    String query = "UPDATE person SET active = ? WHERE id = ?";
+    
+    Connection connection = connectionProvider.connect()
+
+    connection.withCloseable {
+      def statement = connection.prepareStatement(query)
+      statement.setBoolean(1, active)
+      statement.setInt(2, customerId)
+      statement.execute()
+    }
+  }
+  
   /**
    * @inheritDoc
    * @param customerId
    * @param updatedCustomer
    */
   @Override
   void updateCustomer(String customerId, Customer updatedCustomer) {
-    //TODO implement
-    throw new RuntimeException("Method not implemented.")
+
+        int oldCustomerId = Integer.parseInt(customerId)
+    try {
+      if (getCustomer(oldCustomerId)==null) {
+        throw new DatabaseQueryException("Customer is not in the database and can't be updated.")
+      }
+            
+      Connection connection = connectionProvider.connect()
+      connection.setAutoCommit(false)
+
+      connection.withCloseable {it ->
+        try {
+          int newCustomerId = createNewCustomer(it, updatedCustomer)
+          storeAffiliation(it, newCustomerId, updatedCustomer.affiliations)
+          connection.commit()
+          
+          // if our update is successful we set the old customer inactive
+          changeCustomerActiveFlag(oldCustomerId, false)
+          
+        } catch (Exception e) {
+          log.error(e.message)
+          log.error(e.stackTrace.join("\n"))
+          connection.rollback()
+          connection.close()
+          throw new DatabaseQueryException("Could not update customer.")
+        }
+      }
+    } catch (Exception e) {
+      log.error(e)
+      log.error(e.stackTrace.join("\n"))
+      throw new DatabaseQueryException("The customer could not be updated: ${customer.toString()}")
+    }
```

Comment:
```suggestion
      if (getCustomer(oldCustomerId)==null) {
        throw new DatabaseQueryException("Customer is not in the database and can't be updated.")
      }
            
      Connection connection = connectionProvider.connect()
      connection.setAutoCommit(false)

      connection.withCloseable {it ->
        try {
          int newCustomerId = createNewCustomer(it, updatedCustomer)
          storeAffiliation(it, newCustomerId, updatedCustomer.affiliations)
          connection.commit()
          
          // if our update is successful we set the old customer inactive
          changeCustomerActiveFlag(oldCustomerId, false)
          
        } catch (Exception e) {
          log.error(e.message)
          log.error(e.stackTrace.join("\n"))
          connection.rollback()
          connection.close()
          throw new DatabaseQueryException("The customer could not be updated: ${customer.toString()}.")
        }
      }
```

Reply:
You have to be careful here. I think there was a reason the exception was thrown inside the rollback changes.

Reply:
after careful deliberation, it makes sense to me